### ISSUE TITLE
how to apply in Pose Estimation?

### ISSUE THREAD

**Issue by Rookielike** [slryou41/AnchorLoss#4]
thx for your sharing
could you please tell me how to apply anchor loss in pose estimation? 
i try to use MSELoss multiply by anchor_weight which is :
anchor_weight = torch.where(
        torch.eq(targets, 0),
        torch.pow(1 + predictions - pt_pos, gamma),
        torch.full_like(targets,1)
    )
the result is unideal..

**Comment by slryou41** [comment]
Anchor loss is based on BCE not MSE. I haven't applied anchor loss to MSE loss, but it might be interesting to see the result. How did you get the pt_pos? The way I designed the anchor probability is, given the ground truth heatmap, which has Gaussian distribution over spatial dimension, you should get the anchor probability from the prediction by picking up the maximum score at the target area (where the ground truth heatmap value is higher than 0.5). Also, I added a margin variable (-0.05) to pt_pos to penalize the false positive scores which are close to the true positive score.

**Comment by Rookielike** [comment]
@slryou41  thx for your reply
i got pt_pos by the following 
![image](https://user-images.githubusercontent.com/28975635/73705381-c528a900-4730-11ea-923d-dc4f6f221fc9.png)
and i also add a margin variable to pt_pos ,just like this
![image](https://user-images.githubusercontent.com/28975635/73705468-fef9af80-4730-11ea-877a-72cec2de8f2b.png)
but the AP has fallen by 34 %

**Comment by Rookielike** [comment]
@slryou41  i have also tried BSE loss by the following
![image](https://user-images.githubusercontent.com/28975635/73706308-18036000-4733-11ea-8adb-4caab4d60129.png)
i get anchor_probs_neg just like above, and i got 0 AP...

**Comment by slryou41** [comment]
Can you train the model with F.binary_cross_entropy(pt, targets) to see if it works? I was using the gaussian heatmap for BCE and computing all log values (cross entropy computation) in the loss function. Your code seems more optimized than mine... FYI, my implementation was like this:

```
target_mask = (target > 0.0).type('torch.FloatTensor').cuda()
neg_mask = (1 - target_mask)

max_mask = (target > 0.5).type('torch.FloatTensor').cuda()
max_pos = input * max_mask
max_pos = max_pos.max(2)[0].max(2)[0]

max_target = (max_mask * target).max(2)[0].max(2)[0]
pos = Variable(torch.ones((target.size(0), target.size(1)))).cuda() * (1 + self.slack)
pos[max_target > 0] = max_pos[max_target > 0]  # pos == 1 when part annotation does not exist
pos = (pos.view(input.size(0), input.size(1), 1, 1) - self.slack).clamp(min=0).detach()

neg_loss = -(input + (1 - pos)).pow(self.gamma) * torch.log((1 - input).clamp(min=1e-20))
neg_loss = neg_mask * neg_loss

pos_loss = -torch.log(input.clamp(min=1e-20))
bg_loss = -torch.log((1-input).clamp(min=1e-20))

bg_loss = target_mask * bg_loss

loss = (target * pos_loss + (1 - target) * bg_loss) + neg_loss

return loss.mean()
```

**Comment by Rookielike** [comment]
thx for your reply . i will try it later as gpu is used now. Reply to you later.

**Comment by Rookielike** [comment]
hi author,
i have tried the model with F.binary_cross_entropy(pt,targets) and anchor_loss just like your implementation,but the AP in each one was 0.There must be some bug in my code.but when i  trained model using MSEloss ,i got right result.i 'm a little confused.i have two questions as following:
1.how to apply BCEloss in pose estimate? In my way ,generating gaussian heatmap (values are between 0 and 1) and adding nn.Sigmoid() to output and then using BCEloss(using F.binary_cross_entropy(pt,targets) directly or using F.binary_cross_entropy(pt,targets_binary) as targets' values are 0 to 1 not 0 or 1). is my way right?
2. In your code, loss = (target * pos_loss + (1 - target) * bg_loss) + neg_loss .why there is bg_loss in the formula?
could you please give me the answer? thx

**Comment by slryou41** [comment]
I don't think this will fix the bug, but I applied sigmoid in the model file...
1. I think both ways should work. I used the gaussian heatmap (values between 0 and 1).
2. For the area where the ground truth has a value greater than 0, I applied normal BCE loss and only applied AL for the negative area (target == 0). pos loss is ``-log(q)`` and bg_loss is ``-log(1-q)``. ``(target * pos_loss + (1 - target) * bg_loss)`` part is doing normal BCE loss computation. If you use F.binary_cross_entropy, then you don't need to calculate pos_loss and bg_loss.

**Comment by Rookielike** [comment]
thx for your patient reply

**Comment by YuQi9797** [comment]
> Can you train the model with F.binary_cross_entropy(pt, targets) to see if it works? I was using the gaussian heatmap for BCE and computing all log values (cross entropy computation) in the loss function. Your code seems more optimized than mine... FYI, my implementation was like this:
> 
> ```
> target_mask = (target > 0.0).type('torch.FloatTensor').cuda()
> neg_mask = (1 - target_mask)
> 
> max_mask = (target > 0.5).type('torch.FloatTensor').cuda()
> max_pos = input * max_mask
> max_pos = max_pos.max(2)[0].max(2)[0]
> 
> max_target = (max_mask * target).max(2)[0].max(2)[0]
> pos = Variable(torch.ones((target.size(0), target.size(1)))).cuda() * (1 + self.slack)
> pos[max_target > 0] = max_pos[max_target > 0]  # pos == 1 when part annotation does not exist
> pos = (pos.view(input.size(0), input.size(1), 1, 1) - self.slack).clamp(min=0).detach()
> 
> neg_loss = -(input + (1 - pos)).pow(self.gamma) * torch.log((1 - input).clamp(min=1e-20))
> neg_loss = neg_mask * neg_loss
> 
> pos_loss = -torch.log(input.clamp(min=1e-20))
> bg_loss = -torch.log((1-input).clamp(min=1e-20))
> 
> bg_loss = target_mask * bg_loss
> 
> loss = (target * pos_loss + (1 - target) * bg_loss) + neg_loss
> 
> return loss.mean()
> ```

Sir, I writing the loss, like: 
![image](https://user-images.githubusercontent.com/32479497/133879984-eb720748-9a93-4fde-9448-56179e27becb.png)
namely,
![image](https://user-images.githubusercontent.com/32479497/133880051-de71a6a6-e671-47cb-9cfc-7a0d437a0c9f.png)

Does your code mean that?
